chips/rp2040/src/i2c.rs

```diff
@@ -0,0 +1,588 @@
+            if let Some(reason) = self.abort_reason.take() {
+                if reason.matches_all(IC_TX_ABRT_SOURCE::ABRT_7B_ADDR_NOACK::SET) {
+                    Err(hil::i2c::Error::AddressNak)
+                } else if reason.matches_all(IC_TX_ABRT_SOURCE::ABRT_TXDATA_NOACK::SET) {
+                    Err(hil::i2c::Error::DataNak)
+                } else if reason.matches_all(IC_TX_ABRT_SOURCE::ARB_LOST::SET) {
+                    Err(hil::i2c::Error::ArbitrationLost)
+                } else {
+                    Err(hil::i2c::Error::NotSupported)
+                }
+            } else {
+                Ok(())
+            }
+        };
+
+        // Reset state before the callback in case the client wants to start a
+        // new write inside the callback
+        self.state.set(State::Idle);
+
+        self.client.map(|client| match self.buf.take() {
+            None => {}
+            Some(buf) => {
+                client.command_complete(buf, status);
+            }
+        });
+
+        // NOTE:
+        //
+        // The hardware issues a STOP automatically on an abort condition.
+        // Note also the hardware clears RX FIFO as well as TX on abort,
+        // because we set hwparam IC_AVOID_RX_FIFO_FLUSH_ON_TX_ABRT to 0.
+    }
+
+    pub fn handle_interrupt(&self) {
+        match self.state.get() {
+            State::Uninitialized => debug!(
+                "Unexpected IRQ for uninitialized I2C device {}",
+                self.instance_num
+            ),
+            State::Idle => debug!("Unexpected IRQ for idle I2C device {}", self.instance_num),
+            State::WaitingToWriteNextByte => self.write_next_byte(),
+            State::WaitingForStop => self.process_stop_det(),
+        }
+    }
+}
+
+impl hil::i2c::I2CMaster for I2c {
+    fn set_master_client(&self, client: &'static dyn hil::i2c::I2CHwMasterClient) {
+        self.client.set(client);
+    }
+
+    fn enable(&self) {
+        self.enable();
+        // TODO: set as master once we support slave mode too
+    }
+
+    fn disable(&self) {
+        self.disable();
+    }
+
+    fn write_read(
+        &self,
+        _addr: u8,
+        _data: &'static mut [u8],
+        _write_len: u8,
+        _read_len: u8,
+    ) -> Result<(), (hil::i2c::Error, &'static mut [u8])> {
+        todo!()
+    }
+
+    fn write(
+        &self,
+        addr: u8,
+        data: &'static mut [u8],
+        len: u8,
+    ) -> Result<(), (hil::i2c::Error, &'static mut [u8])> {
+        self.buf.put(Some(data));
+
+        if let Err(error) = self.write(addr, len) {
+            // The unwrap should not fail because we have just assigned to buf
+            Err((error, self.buf.take().unwrap()))
+        } else {
+            Ok(())
+        }
+    }
+
+    fn read(
+        &self,
+        _addr: u8,
+        _buffer: &'static mut [u8],
+        _len: u8,
+    ) -> Result<(), (hil::i2c::Error, &'static mut [u8])> {
+        todo!()
+    }
+}
+
+register_structs! {
```

Comment:
Mostly a cosmetic suggestion, to match the structure of the other files, could you move the definitions before the implementations?

Reply:
I prefer it at the bottom because it looks like clutter, but I do prefer consistency more, so sure I'll move it up ASAP.